near-sdk-macros/src/core_impl/info_extractor/attr_sig_info.rs

```diff
@@ -62,103 +169,70 @@ impl AttrSigInfo {
         if let Some(combined_errors) = errors.into_iter().reduce(|mut l, r| (l.combine(r), l).1) {
             return Err(combined_errors);
         }
+
+        // Run early checks to determine the method type
+        let mut visitor: Box<dyn BindgenVisitor> =
+            if original_attrs.iter().any(|a| a.path.to_token_stream().to_string() == "init") {
+                Box::new(InitVisitor::default())
+            } else if Self::is_view(original_sig) {
+                Box::new(ViewVisitor::default())
+            } else {
+                Box::new(CallVisitor::default())
+            };
+
         let ident = original_sig.ident.clone();
         let mut non_bindgen_attrs = vec![];
-        let mut args = vec![];
-        let mut method_type = MethodType::Regular;
-        let mut is_payable = false;
-        let mut is_private = false;
-        let mut is_handles_result = false;
-        // By the default we serialize the result with JSON.
-        let mut result_serializer = SerializerType::JSON;
-
-        let mut payable_attr = None;
+        let mut handles_result = false;
         for attr in original_attrs.iter() {
             let attr_str = attr.path.to_token_stream().to_string();
             match attr_str.as_str() {
                 "init" => {
                     let init_attr: InitAttr = syn::parse2(attr.tokens.clone())?;
-                    if init_attr.ignore_state {
-                        method_type = MethodType::InitIgnoreState;
-                    } else {
-                        method_type = MethodType::Init;
-                    }
+                    visitor.visit_init_attr(attr, &init_attr)?;
                 }
                 "payable" => {
-                    payable_attr = Some(attr);
-                    is_payable = true;
+                    visitor.visit_payable_attr(attr)?;
                 }
                 "private" => {
-                    is_private = true;
+                    visitor.visit_private_attr(attr)?;
                 }
                 "result_serializer" => {
                     let serializer: SerializerAttr = syn::parse2(attr.tokens.clone())?;
-                    result_serializer = serializer.serializer_type;
+                    visitor.visit_result_serializer_attr(attr, &serializer)?;
                 }
                 "handle_result" => {
-                    is_handles_result = true;
+                    handles_result = true;
                 }
```

Comment:
Don't we need `vistor.visit_result` here as well?

Reply:
Nah, we call `visitor.visit_result` below (and pass `handles_result` there) so that it is called both when `handle_result` is present and when it is not. If I move it here, then it won't be called for any of the regular methods.